core/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java

```diff
@@ -45,17 +45,44 @@
     public static XContentBuilder contentBuilder() throws IOException {
         return XContentBuilder.builder(jsonXContent);
     }
-
     private static final JsonFactory jsonFactory;
+
     public static final JsonXContent jsonXContent;
 
+    /*
+     * NOTE: This comment is only meant for maintainers of the Elasticsearch code base and is intentionally not a Javadoc comment as it
+     *       describes an undocumented system property.
+     *
+     *
+     * Determines whether the JSON parser will always check for duplicate keys in JSON content. This behavior is enabled by default but
+     * can be disabled by setting the otherwise undocumented system property "es.json.strict_duplicate_detection" to "false".
+     *
+     * Before we've enabled this mode, we had custom duplicate checks in various parts of the code base. As the user can still disable this
+     * mode and fall back to the legacy duplicate checks, we still need to keep the custom duplicate checks around and we also need to keep
+     * the tests around.
+     *
+     * If this fallback via system property is removed one day in the future you can remove all tests that call this method and also remove
+     * the corresponding custom duplicate check code.
+     *
+     */
+    public static boolean isStrictDuplicateDetectionEnabled() {
+        // Don't allow duplicate keys in JSON content by default but let the user opt out
+        return Boolean.valueOf(System.getProperty("es.json.strict_duplicate_detection", "true"));
+    }
+
     static {
         jsonFactory = new JsonFactory();
         jsonFactory.configure(JsonGenerator.Feature.QUOTE_FIELD_NAMES, true);
         jsonFactory.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
         jsonFactory.configure(JsonFactory.Feature.FAIL_ON_SYMBOL_HASH_OVERFLOW, false); // this trips on many mappings now...
         // Do not automatically close unclosed objects/arrays in com.fasterxml.jackson.core.json.UTF8JsonGenerator#close() method
         jsonFactory.configure(JsonGenerator.Feature.AUTO_CLOSE_JSON_CONTENT, false);
+        try {
+            jsonFactory.configure(JsonParser.Feature.STRICT_DUPLICATE_DETECTION, isStrictDuplicateDetectionEnabled());
+        } catch (Exception ex) {
```

Comment:
Yes it is. A user could still pass `-Des.json.strict_duplicate_detection=complete_bogus` and provoke an exception when we try to convert the system property to a `boolean`. I don't want to be lenient and terminate early then.